src/main/java/org/hashids/Hashids.java

```diff
@@ -23,154 +29,164 @@
   private static final double SEP_DIV = 3.5;
   private static final int GUARD_DIV = 12;
+  private static final Pattern WORD_PATTERN = Pattern.compile("[\\w\\W]{1,12}");
 
-  private final String salt;
+  private final char[] salt;
   private final int minHashLength;
-  private final String alphabet;
-  private final String seps;
-  private final String guards;
+  private final char[] alphabet;
+  private final char[] seps;
+  private final char[] guards;
+  private final String guardsRegExp;
+  private final String sepsRegExp;
+  private final char[] validChars;
 
   public Hashids() {
-    this(DEFAULT_SALT);
+    this(DEFAULT_SALT, DEFAULT_MIN_HASH_LENGTH, DEFAULT_ALPHABET);
   }
 
   public Hashids(String salt) {
-    this(salt, 0);
+    this(salt, DEFAULT_MIN_HASH_LENGTH);
   }
 
   public Hashids(String salt, int minHashLength) {
-    this(salt, minHashLength, DEFAULT_ALPHABET);
+    this((salt == null) ? null : salt.toCharArray(), minHashLength, DEFAULT_ALPHABET);
   }
 
   public Hashids(String salt, int minHashLength, String alphabet) {
-    this.salt = salt != null ? salt : DEFAULT_SALT;
-    this.minHashLength = minHashLength > 0 ? minHashLength : DEFAULT_MIN_HASH_LENGTH;
+    this((salt == null) ? null : salt.toCharArray(),
+        minHashLength,
+        (alphabet == null) ? null : alphabet.toCharArray());
+  }
 
-    final StringBuilder uniqueAlphabet = new StringBuilder();
-    for (int i = 0; i < alphabet.length(); i++) {
-      if (uniqueAlphabet.indexOf(String.valueOf(alphabet.charAt(i))) == -1) {
-        uniqueAlphabet.append(alphabet.charAt(i));
-      }
-    }
+  private Hashids(char[] salt, int minHashLength, char[] alphabet) {
+    assert salt != null;
+    assert minHashLength >= 0;
+    assert alphabet != null;
 
-    alphabet = uniqueAlphabet.toString();
+    this.salt = salt;
+    this.minHashLength = minHashLength;
 
-    if (alphabet.length() < MIN_ALPHABET_LENGTH) {
-      throw new IllegalArgumentException(
-          "alphabet must contain at least " + MIN_ALPHABET_LENGTH + " unique characters");
-    }
-
-    if (alphabet.contains(" ")) {
-      throw new IllegalArgumentException("alphabet cannot contains spaces");
-    }
+    // alphabet
+    validateAlphabet(alphabet);
 
     // seps should contain only characters present in alphabet;
     // alphabet should not contains seps
-    String seps = DEFAULT_SEPS;
-    for (int i = 0; i < seps.length(); i++) {
-      final int j = alphabet.indexOf(seps.charAt(i));
-      if (j == -1) {
-        seps = seps.substring(0, i) + " " + seps.substring(i + 1);
-      } else {
-        alphabet = alphabet.substring(0, j) + " " + alphabet.substring(j + 1);
-      }
-    }
+    char[] seps = cleanup(DEFAULT_SEPS, alphabet);
+    alphabet = removeAll(alphabet, seps);
 
-    alphabet = alphabet.replaceAll("\\s+", "");
-    seps = seps.replaceAll("\\s+", "");
-    seps = Hashids.consistentShuffle(seps, this.salt);
+    seps = Hashids.consistentShuffle(seps, salt);
```

Comment:
always prefix methods/attributes with `this.`.

Reply:
Why? This is not a JAVA rule. Perhaps a JS one.